src/models/WidgetTemplateImport.php

```diff
@@ -0,0 +1,279 @@
+                } else {
+                    $this->_filenames[] = $filename;
+                }
+            }
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @return array
+     */
+    public function attributeLabels()
+    {
+        $attributeLabels = parent::attributeLabels();
+        $attributeLabels['tarFiles'] = \Yii::t('widgets', 'Files');
+        return $attributeLabels;
+    }
+
+    /**
+     * @param string $filepath
+     * @param string $extractDirectory
+     * @return bool
+     */
+    protected function extractTar(string $filepath, string $extractDirectory): bool
+    {
+        return (new \PharData($filepath))->extractTo($extractDirectory);
+    }
+
+    /**
+     * @return bool
+     */
+    protected function extractFiles(): bool
+    {
+        foreach ($this->_filenames as $filename) {
+            if ($this->extractTar($filename, $this->getExtractDirectory($filename)) === false) {
+                $this->addError('tarFiles', \Yii::t('widgets', 'Error while extracting file'));
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * @param string $filename
+     * @return string
+     */
+    protected function getExtractDirectory(string $filename): string
+    {
+        return $this->_importDirectory . DIRECTORY_SEPARATOR . Inflector::slug(basename($filename));
+    }
+
+    /**
+     * @return bool
+     */
+    protected function import(): bool
+    {
+        $transaction = WidgetTemplate::getDb()->beginTransaction();
+        try {
+            if ($transaction && $transaction->getIsActive()) {
+                foreach ($this->_filenames as $filename) {
+                    if (! $this->importTemplateByDirectory($filename)) {
+                        throw new \Exception('Error while importing files');
+                    }
+                }
+            }
+            $transaction->commit();
+            return true;
+        } catch (\Exception $e) {
+            \Yii::error($e->getMessage());
+            $this->addError('tarFiles', $e->getMessage());
+        }
+        $transaction->rollBack();
+        return false;
+    }
+
+    /**
+     * @param string $filename
+     * @return bool
+     */
+    protected function importTemplateByDirectory(string $filename): bool
+    {
+        $extractDirectory = $this->getExtractDirectory($filename);
+        $meta = $this->templateMeta($extractDirectory);
+        $template = $this->templateContent($extractDirectory);
+        $schema = $this->schemaContent($extractDirectory);
+
+        $model = new WidgetTemplate([
+            'name' => $meta['name'],
+            'php_class' => $meta['php_class'],
+            'twig_template' => $template,
+            'json_schema' => $schema
+        ]);
+
+        if ($model->save() === false) {
+            $this->addError('tarFiles', print_r($model->getErrors(), true));
```

Comment:
Maybe would c|should format error messages a little bit ;-)
![Screenshot 2022-05-10 at 10 14 53](https://user-images.githubusercontent.com/825574/167582261-54dd9c8d-5f8e-4d60-8377-329eb50b69fc.png)